sdcm/cluster.py

```diff
@@ -973,8 +974,8 @@ def stop_task_threads(self, timeout=10):
         if self._scylla_manager_journal_thread:
             self.stop_scylla_manager_log_capture(timeout)
         if self._decoding_backtraces_thread:
-            Setup.DECODING_QUEUE.put(None)
             Setup.DECODING_QUEUE.join()
+            Setup.DECODING_QUEUE.put(None)
             self._decoding_backtraces_thread.join(timeout)
             self._decoding_backtraces_thread = None
```

Comment:
I think this will solve the issue. At some time we will get to the same situation. Currently we got race between stopping decoding_backtrace_thread and DECODING_QUEUE. Problem looks that decoding thread is stopped first by condition termintation_event and Decoding_QUEUE empty.if this happened first and then putting None object, then we got stacking on DECODING.join().
in this fix, we put to Setup.Decoding_Queue after join(), in this case  Decoding queue already empty and bactrace_thread could already stopped, and we get situation the Decoding queue will continue to run.

I think better solution, not to put None at all. in this case when the decoding queue become empty and termination event will set, then DECODING_queue.join() will not hang and queue will continue to run, and decoding bactrace thread will be stopped.

Reply:
> I think this will solve the issue. At some time we will get to the same situation. Currently we got race between stopping decoding_backtrace_thread and DECODING_QUEUE. Problem looks that decoding thread is stopped first by condition termintation_event and Decoding_QUEUE empty.if this happened first and then putting None object, then we got stacking on DECODING.join().

That's right

> in this fix, we put to Setup.Decoding_Queue after join(), in this case Decoding queue already empty and bactrace_thread could already stopped, and we get situation the Decoding queue will continue to run.

Yes.

> I think better solution, not to put None at all. in this case when the decoding queue become empty and termination event will set, then DECODING_queue.join() will not hang and queue will continue to run, and decoding bactrace thread will be stopped.

We can't put(None) as past, it's not necessary to put(None) after join().
I will retest by removing the put(None).  If it works I will update the PR.

Reply:
The test is actually passed. I will update the PR to remove put(None)

- https://jenkins.scylladb.com/job/scylla-staging/job/amos/job/rolling-upgrade-3.2/job/rolling-upgrade-centos7/9/console